internal/provider/iam/resource_group.go

```diff
@@ -0,0 +1,228 @@
+package iam
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-iam/stable/2019-12-10/client/groups_service"
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-iam/stable/2019-12-10/models"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	clients "github.com/hashicorp/terraform-provider-hcp/internal/clients"
+	"github.com/hashicorp/terraform-provider-hcp/internal/hcpvalidator"
+)
+
+func NewGroupResource() resource.Resource {
+	return &resourceGroup{}
+}
+
+type resourceGroup struct {
+	client *clients.Client
+}
+
+func (r *resourceGroup) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_group"
+}
+
+func (r *resourceGroup) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: fmt.Sprintf(`The group resource manages a HCP Group.
+
+The user or service account that is running Terraform when creating an %s resource must have %s on the parent resource; either the project or organization.`,
+			"`hcp_group`", "`roles/admin`"),
+		Attributes: map[string]schema.Attribute{
+			"resource_id": schema.StringAttribute{
+				Computed:    true,
+				Description: "The group's unique identifier",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"resource_name": schema.StringAttribute{
+				Computed: true,
+				Description: fmt.Sprintf("The group's resource name in the format `%s` or `%s`",
+					"iam/organization/<organization_id>/group/<name>", "<name>"),
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"display_name": schema.StringAttribute{
+				Required:    true,
+				Description: "The group's display_name.",
+				Validators: []validator.String{
+					hcpvalidator.ResourceNamePart(),
+					stringvalidator.LengthBetween(3, 90),
+				},
+			},
+			"description": schema.StringAttribute{
+				Description: "The group's description ",
+				Optional:    true,
+				Validators: []validator.String{
+					hcpvalidator.ResourceNamePart(),
```

Comment:
**question:** shouldn't we allow characters not valid for a resource name?

Reply:
Hmm, good question @squaresurf! In the backend, currently we allow those characters and then convert them to a valid resource name. I went with this validator since that was the convention established in the service principal and workload identity resources, but I'm open to removing this constraint given we handle conversions in the backend for the groups handler.